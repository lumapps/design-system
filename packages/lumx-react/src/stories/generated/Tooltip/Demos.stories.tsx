/**
 * File generated when storybook is started. Do not edit directly!
 */
export default { title: 'LumX components/tooltip/Demos' };

export { App as Default } from './default';
export { App as Delay } from './delay';
export { App as Placement } from './placement';
