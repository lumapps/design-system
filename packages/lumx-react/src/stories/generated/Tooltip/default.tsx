/home/gcornut/git/design-system/packages/site-demo/content/product/components/tooltip/react/default.tsx