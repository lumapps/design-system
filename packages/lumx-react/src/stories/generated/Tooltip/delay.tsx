/home/gcornut/git/dup-design-system/packages/site-demo/content/product/components/tooltip/react/delay.tsx