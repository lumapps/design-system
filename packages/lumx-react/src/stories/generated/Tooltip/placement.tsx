/home/gcornut/git/dup-design-system/packages/site-demo/content/product/components/tooltip/react/placement.tsx