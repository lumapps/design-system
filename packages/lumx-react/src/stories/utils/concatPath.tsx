export type PathPart = string | number;
export type Path = PathPart | Array<PathPart>;

/**
 * Concat flatten object path
 *
 * @example concatPath('foo', 'bar') // => 'foo.bar'
 * @example concatPath(['foo', 0]) // => 'foo[0]'
 * @example concatPath('foo', 0, ['bar']) // => 'foo[0].bar'
 */
export const concatPath = (...prefix: Path[]) => {
    const [first, ...rest] = prefix.flat();
    return rest.reduce<string>((acc, part) => {
        if (typeof part === 'number') return `${acc}[${part}]`;
        return `${acc}.${part}`;
    }, String(first));
};
