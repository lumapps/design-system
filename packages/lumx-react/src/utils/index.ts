/**
 * WARNING: All modules exported here are exposed to NPM in '@lumx/react/utils'.
 */

export { ClickAwayProvider } from './ClickAwayProvider';
export { Portal, type PortalProps, type PortalInit, PortalProvider, type PortalProviderProps } from './Portal';
