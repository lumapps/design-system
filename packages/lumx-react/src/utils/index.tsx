export * from '@lumx/core/js/utils';

export * from './getRootClassName';

export * from './htmlDecode';

export * from './partitionMulti';

export * from './type';
