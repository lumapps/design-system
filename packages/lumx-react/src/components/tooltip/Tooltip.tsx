/* eslint-disable react-hooks/rules-of-hooks */
import React, { forwardRef, ReactNode, useMemo, useState } from 'react';
import { createPortal } from 'react-dom';
import { usePopper } from 'react-popper';
import { uid } from 'uid';

import classNames from 'classnames';

import { Placement } from '@lumx/react/components/popover/Popover';

import { DOCUMENT } from '@lumx/react/constants';

import { Comp, GenericProps, getRootClassName, handleBasicClasses } from '@lumx/react/utils';
import { mergeRefs } from '@lumx/react/utils/mergeRefs';

import { useInjectTooltipRef } from './useInjectTooltipRef';
import { useTooltipOpen } from './useTooltipOpen';

/** Position of the tooltip relative to the anchor element. */
export type TooltipPlacement = Extract<Placement, 'top' | 'right' | 'bottom' | 'left'>;

/**
 * Defines the props of the component.
 */
export interface TooltipProps extends GenericProps {
    /** Anchor (element on which we activate the tooltip). */
    children: ReactNode;
    /** Delay (in ms) before closing the tooltip. */
    delay?: number;
    /** Whether the tooltip is displayed even without the mouse hovering the anchor. */
    forceOpen?: boolean;
    /** Label text. */
    label?: string | null | false;
    /** Placement of the tooltip relative to the anchor. */
    placement?: TooltipPlacement;
}

/**
 * Component display name.
 */
const COMPONENT_NAME = 'Tooltip';

/**
 * Component default class name and class prefix.
 */
const CLASSNAME = getRootClassName(COMPONENT_NAME);

/**
 * Component default props.
 */
const DEFAULT_PROPS: Partial<TooltipProps> = {
    delay: 500,
    placement: Placement.BOTTOM,
};

/**
 * Arrow size (in pixel).
 */
const ARROW_SIZE = 8;

/**
 * Tooltip component.
 *
 * @param  props Component props.
 * @param  ref   Component ref.
 * @return React element.
 */
export const Tooltip: Comp<TooltipProps, HTMLDivElement> = forwardRef((props, ref) => {
    if (!DOCUMENT) {
        // Can't render in SSR.
        return null;
    }
    const { label, children, className, delay, placement, forceOpen, ...forwardedProps } = props;
    if (!label) {
        return <>{children}</>;
    }

    const id = useMemo(() => `tooltip-${uid()}`, []);

    const [popperElement, setPopperElement] = useState<null | HTMLElement>(null);
    const [anchorElement, setAnchorElement] = useState<null | HTMLElement>(null);
    const { styles, attributes } = usePopper(anchorElement, popperElement, {
        placement,
        modifiers: [
            {
                name: 'offset',
                options: { offset: [0, ARROW_SIZE] },
            },
        ],
    });

    const position = attributes?.popper?.['data-popper-placement'] ?? placement;
    const isOpen = useTooltipOpen(delay as number, anchorElement) || forceOpen;
    const wrappedChildren = useInjectTooltipRef(children, setAnchorElement, isOpen as boolean, id);

    return (
        <>
            {wrappedChildren}
            {isOpen &&
                createPortal(
                    <div
                        ref={mergeRefs(ref, setPopperElement)}
                        {...forwardedProps}
                        id={id}
                        role="tooltip"
                        aria-label={label}
                        className={classNames(className, handleBasicClasses({ prefix: CLASSNAME, position }))}
                        style={styles.popper}
                        {...attributes.popper}
                    >
                        <div className={`${CLASSNAME}__arrow`} />
                        <div className={`${CLASSNAME}__inner`}>
                            {label.indexOf('\n') !== -1
                                ? label.split('\n').map((sentence: string) => <p key={sentence}>{sentence}</p>)
                                : label}
                        </div>
                    </div>,
                    document.body,
                )}
        </>
    );
});
Tooltip.displayName = COMPONENT_NAME;
Tooltip.className = CLASSNAME;
Tooltip.defaultProps = DEFAULT_PROPS;
