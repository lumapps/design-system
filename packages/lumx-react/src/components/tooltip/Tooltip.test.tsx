import React, { ReactElement } from 'react';

import { mount, shallow } from 'enzyme';
import 'jest-enzyme';

import { Wrapper, commonTestsSuite } from '@lumx/react/testing/utils';

import { Button, Icon } from '@lumx/react';
import { Tooltip, TooltipProps } from './Tooltip';

const CLASSNAME = Tooltip.className as string;

jest.mock('uid', () => ({ uid: () => 'uid' }));

type SetupProps = Partial<TooltipProps>;

/**
 * Mounts the component and returns common DOM elements / data needed in multiple tests further down.
 */
const setup = (propsOverride: SetupProps = {}, shallowRendering = true) => {
    const props: any = { ...propsOverride };
    const renderer: (el: ReactElement) => Wrapper = shallowRendering ? shallow : mount;
    const wrapper = renderer(
        <Tooltip label="Tooltip" {...props}>
            Anchor
        </Tooltip>,
    );
    const tooltip = wrapper.find(`.${CLASSNAME}`);

    return {
        props,
        tooltip,
        wrapper,
    };
};

jest.mock('./useTooltipOpen', () => ({ useTooltipOpen: () => true }));

describe(`<${Tooltip.displayName}>`, () => {
    // 1. Test render via snapshot (default states of component).
    describe('Snapshots and structure', () => {
        // Here is an example of a basic rendering check, with snapshot.

        it('should render correctly', () => {
            const { tooltip, wrapper } = setup();
            expect(wrapper).toMatchSnapshot();

            expect(tooltip).toExist();
            expect(tooltip).toHaveClassName(CLASSNAME);
        });

        it('should return children when empty label', () => {
            const wrapper = shallow(
                <Tooltip label="">
                    <Icon icon="icon" />
                </Tooltip>,
            );
            expect(wrapper).toMatchSnapshot();
        });

        it('should wrap children', () => {
            // Wrap string children
            const wrapper1 = shallow(<Tooltip label="tooltip1">children</Tooltip>);
            expect(wrapper1).toMatchSnapshot();

            // Wrap fragment children
            const wrapper2 = shallow(
                <Tooltip label="tooltip1">
                    <>children</>
                </Tooltip>,
            );
            expect(wrapper2).toMatchSnapshot();
        });

        it('should not wrap Icon', () => {
            const wrapper = shallow(
                <Tooltip label="tooltip1">
                    <Icon icon="icon" />
                </Tooltip>,
            );
            expect(wrapper).toMatchSnapshot();
        });

        it('should not wrap Button', () => {
            const wrapper = shallow(
                <Tooltip label="tooltip1">
                    <Button>button</Button>
                </Tooltip>,
            );
            expect(wrapper).toMatchSnapshot();
        });

        it('should work on disabled elements', () => {
            const wrapper = shallow(
                <Tooltip label="Tooltip on disabled button" forceOpen>
                    <Button isDisabled>Empty</Button>
                </Tooltip>,
            );
            expect(wrapper).toMatchSnapshot();
        });
    });

    // Common tests suite.
    commonTestsSuite(setup, { className: 'tooltip', prop: 'tooltip' }, { className: CLASSNAME });
});
