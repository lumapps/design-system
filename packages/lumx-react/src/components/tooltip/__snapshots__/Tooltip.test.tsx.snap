// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tooltip> Snapshots and structure should not wrap Button 1`] = `
<Fragment>
  <Button
    aria-describedby="tooltip-uid"
    emphasis="high"
    size="m"
    theme="light"
  >
    button
  </Button>
  <Portal
    containerInfo={<body />}
  >
    <div
      aria-label="tooltip1"
      className="lumx-tooltip lumx-tooltip--position-bottom"
      id="tooltip-uid"
      role="tooltip"
      style={
        Object {
          "left": "0",
          "position": "absolute",
          "top": "0",
        }
      }
    >
      <div
        className="lumx-tooltip__arrow"
      />
      <div
        className="lumx-tooltip__inner"
      >
        tooltip1
      </div>
    </div>
  </Portal>
</Fragment>
`;

exports[`<Tooltip> Snapshots and structure should not wrap Icon 1`] = `
<Fragment>
  <Icon
    aria-describedby="tooltip-uid"
    icon="icon"
  />
  <Portal
    containerInfo={<body />}
  >
    <div
      aria-label="tooltip1"
      className="lumx-tooltip lumx-tooltip--position-bottom"
      id="tooltip-uid"
      role="tooltip"
      style={
        Object {
          "left": "0",
          "position": "absolute",
          "top": "0",
        }
      }
    >
      <div
        className="lumx-tooltip__arrow"
      />
      <div
        className="lumx-tooltip__inner"
      >
        tooltip1
      </div>
    </div>
  </Portal>
</Fragment>
`;

exports[`<Tooltip> Snapshots and structure should render correctly 1`] = `
<Fragment>
  <div
    aria-describedby="tooltip-uid"
    className="lumx-tooltip-anchor-wrapper"
  >
    Anchor
  </div>
  <Portal
    containerInfo={<body />}
  >
    <div
      aria-label="Tooltip"
      className="lumx-tooltip lumx-tooltip--position-bottom"
      id="tooltip-uid"
      role="tooltip"
      style={
        Object {
          "left": "0",
          "position": "absolute",
          "top": "0",
        }
      }
    >
      <div
        className="lumx-tooltip__arrow"
      />
      <div
        className="lumx-tooltip__inner"
      >
        Tooltip
      </div>
    </div>
  </Portal>
</Fragment>
`;

exports[`<Tooltip> Snapshots and structure should return children when empty label 1`] = `
<Fragment>
  <Icon
    icon="icon"
  />
</Fragment>
`;

exports[`<Tooltip> Snapshots and structure should work on disabled elements 1`] = `
<Fragment>
  <div
    aria-describedby="tooltip-uid"
    className="lumx-tooltip-anchor-wrapper"
  >
    <Button
      emphasis="high"
      isDisabled={true}
      size="m"
      theme="light"
    >
      Empty
    </Button>
  </div>
  <Portal
    containerInfo={<body />}
  >
    <div
      aria-label="Tooltip on disabled button"
      className="lumx-tooltip lumx-tooltip--position-bottom"
      id="tooltip-uid"
      role="tooltip"
      style={
        Object {
          "left": "0",
          "position": "absolute",
          "top": "0",
        }
      }
    >
      <div
        className="lumx-tooltip__arrow"
      />
      <div
        className="lumx-tooltip__inner"
      >
        Tooltip on disabled button
      </div>
    </div>
  </Portal>
</Fragment>
`;

exports[`<Tooltip> Snapshots and structure should wrap children 1`] = `
<Fragment>
  <div
    aria-describedby="tooltip-uid"
    className="lumx-tooltip-anchor-wrapper"
  >
    children
  </div>
  <Portal
    containerInfo={<body />}
  >
    <div
      aria-label="tooltip1"
      className="lumx-tooltip lumx-tooltip--position-bottom"
      id="tooltip-uid"
      role="tooltip"
      style={
        Object {
          "left": "0",
          "position": "absolute",
          "top": "0",
        }
      }
    >
      <div
        className="lumx-tooltip__arrow"
      />
      <div
        className="lumx-tooltip__inner"
      >
        tooltip1
      </div>
    </div>
  </Portal>
</Fragment>
`;

exports[`<Tooltip> Snapshots and structure should wrap children 2`] = `
<Fragment>
  children
  <Portal
    containerInfo={<body />}
  >
    <div
      aria-label="tooltip1"
      className="lumx-tooltip lumx-tooltip--position-bottom"
      id="tooltip-uid"
      role="tooltip"
      style={
        Object {
          "left": "0",
          "position": "absolute",
          "top": "0",
        }
      }
    >
      <div
        className="lumx-tooltip__arrow"
      />
      <div
        className="lumx-tooltip__inner"
      >
        tooltip1
      </div>
    </div>
  </Portal>
</Fragment>
`;
